File: src/medium/WordBreak.java
package medium;

import java.util.Set;

public class WordBreak {
    public boolean wordBreak(String s, Set<String> wordDict) {
        if (s == null || s.length() == 0 || wordDict.size() == 0) {
            return false;
        }
        boolean[] memo = new boolean[s.length() + 1];
        memo[0] = true;
        // i is NOT the index of the string, it is the number count of the string
        // it then is the memo index, and end position of substring method
        for (int i = 1; i <= s.length(); i++) {
            // j is the index of the string
            for (int j = i - 1; j >= 0; j--) {
                // if p1: 0 -> j - 1 is true
                // p1 is store in memo[j]
                // then check p2: j -> i
                if (memo[j] && wordDict.contains(s.substring(j, i))) {
                    memo[i] = true;
                    break;
                }
            }
        }
        return memo[s.length()];
    }
}

File: csharp/LeetCode/Medium/WordBreak.cs
namespace LeetCode.Medium.WordBreak
{
    using System.Collections.Generic;

    public class Solution
    {
        private readonly Dictionary<string, bool> Memo = new Dictionary<string, bool>();
        public bool WordBreak(string s, IList<string> wordDict)
        {
            var hs = new HashSet<string>(wordDict);
            return DFS(s, hs);
        }

        bool DFS(string s, HashSet<string> hs)
        {
            if (Memo.ContainsKey(s))
            {
                return Memo[s];
            }

            if (hs.Contains(s))
            {
                return true;
            }

            for (var i = 1; i < s.Length; i++)
            {
                var left = s.Substring(0, i);
                if (!hs.Contains(left))
                {
                    continue;
                }
                var right = s.Substring(i);
                var hasRight = DFS(right, hs);
                Memo[right] = hasRight;
                if (hasRight)
                {
                    return true;
                }
            }
            Memo[s] = false;
            return false;
        }
    }
}
